### PR TITLE
Improve generation, shrinking and showing for the `Stream` type.

### PR DESCRIPTION
**Pull request by jorisdral**
Some things I noticed when looking into the `Errors` and `Stream` module:
* `Stream`s can be possibly infinite, but in reality we only use finite `Stream`s for `Error`s.
* Shrinking diverges for infinite `Stream`s.
* When we `show` a `Stream`, we just `show` a prefix of the `Stream`. 

This PR updates `Stream` to behave much like [`Test.QuickCheck.InfiniteList`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#t:InfiniteList) in terms of generation, shrinking and showing. Furthermore, we clean up the module and establish a (safe/unsafe) interface to `Stream`s.

### CODE REVIEW

**Comment by jorisdral** [comment]
Current dependencies on/for this PR:
* main
  * **PR #26** <a href="https://app.graphite.dev/github/pr/input-output-hk/fs-sim/26" target="_blank">Graphite</a> 
    * **PR #27** <a href="https://app.graphite.dev/github/pr/input-output-hk/fs-sim/27" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/input-output-hk/fs-sim/27?utm_source=stack-comment).

**Comment by jorisdral** [comment]
This PR should add a changelog entry

File: fs-sim/src/System/FS/Sim/Stream.hs [hunk 1]
```diff
@@ -20,55 +32,137 @@ import           Test.QuickCheck (Gen)
...
 -- A stream consisting of only 'Nothing's (even if it is only one) is not
 -- considered to be empty.
 null :: Stream a -> Bool
-null (Stream []) = True
-null _           = False
+null (Stream _ []) = True
+null _             = False
+
+{-------------------------------------------------------------------------------
+  Generation and shrinking
```
Comment (jorisdral) [inline fs-sim/src/System/FS/Sim/Stream.hs:116]: We should look into whether it makes sense to test these generators and this shrinker.

File: fs-sim/src/System/FS/Sim/Error.hs [hunk 2]
```diff
@@ -66,7 +67,8 @@ import           System.FS.API.Types
 
 import           System.FS.Sim.MockFS (HandleMock, MockFS)
 import qualified System.FS.Sim.STM as Sim
-import           System.FS.Sim.Stream
+import           System.FS.Sim.Stream (Stream)
+import qualified System.FS.Sim.Stream as Stream
```
Comment (jorisdral) [inline fs-sim/src/System/FS/Sim/Error.hs:71]: `System.FS.Sim.Stream` is re-exported by the `Error` module, and qualified imports are not re-exported

**Review by jasagredo — APPROVED** [review]
This looks good. I'm not familiar with Stream and how we use it, but the functions you defined and adapted look reasonable

File: fs-sim/src/System/FS/Sim/Error.hs (hunk 2 above)
Comment (jorisdral) [reply fs-sim/src/System/FS/Sim/Error.hs]: No longer relevant

**Pull request merged**